sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.CosmosAsyncClient
+import com.azure.cosmos.models.CosmosQueryRequestOptions
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ArrayNode, BinaryNode, BooleanNode, DecimalNode, DoubleNode,
+    FloatNode, IntNode, LongNode, NullNode, ObjectNode, TextNode}
+
+
+// scalastyle:off underscore.import
+import org.apache.spark.sql.types._
+import scala.collection.JavaConverters._
+// scalastyle:on underscore.import
+
+// Infers a schema by reading sample data from a source container.
+private object CosmosTableSchemaInferer
+    extends CosmosLoggingTrait {
+    val RAW_JSON_BODY_ATTRIBUTE_NAME = "_rawBody"
+    private val TIMESTAMP_ATTRIBUTE_NAME = "_ts"
+    private val ID_ATTRIBUTE_NAME = "id"
+    private val defaultSchemaForInferenceDisabled = StructType(Seq(
+        StructField(RAW_JSON_BODY_ATTRIBUTE_NAME, StringType),
+        StructField(ID_ATTRIBUTE_NAME, StringType),
+        StructField(TIMESTAMP_ATTRIBUTE_NAME, LongType)
+    ))
+
+    def inferSchema(inferredItems : Seq[ObjectNode]): StructType = {
+        if (inferredItems.isEmpty){
+            // No documents to infer from
+            StructType(Seq())
+        }
+
+        // Create a unique map of all distinct properties from documents
+        // If 2 documents contain the same property name but different type, the last scanned one would define the final
+        // type
+        val uniqueStructFields = inferredItems.foldLeft(Map.empty[String, StructField])({
+            case (map, item) => inferDataTypeFromObjectNode(item) match {
+                case Some(mappedList) => map ++ mappedList
+            }
+        })
+
+        StructType(uniqueStructFields.valuesIterator.toSeq)
+    }
+
+    def inferSchema(client: CosmosAsyncClient,
+                    userConfig: Map[String, String]): StructType = {
+        val cosmosReadConfig = CosmosReadConfig.parseCosmosReadConfig(userConfig)
+        if (cosmosReadConfig.inferSchemaEnabled) {
+            val cosmosContainerConfig = CosmosContainerConfig.parseCosmosContainerConfig(userConfig)
+            val sourceContainer = client.getDatabase(cosmosContainerConfig.database).getContainer(cosmosContainerConfig.container)
+
+            val queryOptions = new CosmosQueryRequestOptions()
+            queryOptions.setMaxBufferedItemCount(cosmosReadConfig.inferSchemaSamplingSize)
+            val queryText = s"select TOP ${cosmosReadConfig.inferSchemaSamplingSize} * from c"
```

Comment:
Not blocking - just something to discuss I want to track here: 
This approach will make it very hard for customers to use schema inference if they store documents with multiple schemas in one container - like one common pattern for IoT scenarios is that for each device you store a summary (basically the digital twin with the "materialized view" of this device you need most commonly for the app/UI) and the raw events (usually with TTL of a couple of days) in the same container (to be able to insert event and update view in an atomic transaction). So I am wondering whether we can either pipe some of the filtering/query from Spark into the schema inference here (not sure whether you even have access to the query context yet) or whether we should allow to specify filtering conditions/query to be used for schema inference in the config options?

Reply:
I checked other Spark integrations and they follow a similar approach (sampling documents), the thing is, with a NoSQL store, it's is rather complicated to infer the schema, like you said, there could be documents with varied schema. We could potentially add a config to inject a custom query?

Reply:
Yes- I like the approach of having a separate optional config property to define the sampling query. Let's discuss with Mo - would be separate PR anyway.

Reply:
I think adding a custom Query will be very helpful. yes. can go in a separate PR though.

Reply:
Will do it in a following PR